### PR TITLE
feat: support reading configuration form xds(mvp)

### PR DESCRIPTION
**Pull request by tzssangglass**
### Description

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added in this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backwards compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: t/amesh-agent/config_shdict.t [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  config_center: shdict
+  enable_admin: false
+--- config
+    location /t {
+        content_by_lua_block {
+            ngx.say("ok")
+        }
+    }
+--- no_error_log eval
+qr/can not load amesh agent/
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply t/amesh-agent/config_shdict.t]: fixed

File: apisix/core/config_shdict.lua [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+local router_config     = ngx.shared["router-config"]
+
+local process
+if is_http then
+    process = require("ngx.process")
+end
+
+
+ffi.cdef[[
+extern void CreateMock(void* writeRoute);
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/core/config_shdict.lua]: `init()` is the name of a reserved function of go, with special meaning, I replaced it with `initial()`

File: apisix/core/config_shdict.lua [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+local base              = require("resty.core.base")
+local config_local      = require("apisix.core.config_local")
+local table             = table
+local is_http           = ngx.config.subsystem == "http"
+local string            = string
+local io                = io
+local package           = package
+local new_tab           = base.new_tab
+local ngx_timer_at      = ngx.timer.at
+local ffi               = require 'ffi'
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/core/config_shdict.lua]: fixed

File: apisix/core/config_shdict.lua [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        local router_zone = C.ngx_http_lua_ffi_shdict_udata_to_zone(router_config[1])
+        local router_shd_cdata = ffi.cast("void*", router_zone)
+        ameshagent.CreateMock(router_shd_cdata)
+    end)
+end
+
+
+
+function _M.init_worker()
+    local lib_name = "libameshagent"
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/core/config_shdict.lua]: fixed

File: apisix/core/config_shdict.lua [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        local ameshagent, tried_paths = load_shared_lib(lib_name)
+
+        if not ameshagent then
+            tried_paths[#tried_paths + 1] = 'tried above paths but can not load '
+                    .. lib_name
+            error("can not load amesh agent, tried paths: " ..
+                    table.concat(tried_paths, '\r\n', 1, #tried_paths))
+        end
+
+        local router_zone = C.ngx_http_lua_ffi_shdict_udata_to_zone(router_config[1])
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/core/config_shdict.lua:96]: `router_config` type is `table`
`router_config[1]` type is `userdata`
`router_zone ` type is `cdata`

use `ngx_http_lua_ffi_shdict_udata_to_zone` here it is for passing pointer variables(`cdata`) to Amesh

File: apisix/core/config_shdict.lua [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+local router_config     = ngx.shared["router-config"]
+
+local process
+if is_http then
+    process = require("ngx.process")
+end
+
+
+ffi.cdef[[
+extern void initial(void* writeRoute);
```
Comment (membphis) [inline apisix/core/config_shdict.lua:43]: we should use standard C style: `write_route `

File: apisix/core/config_shdict.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (membphis) [reply apisix/core/config_shdict.lua:96]: ok, got it

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -29,7 +29,7 @@ jobs:
         test_dir:
           - t/plugin
           - t/admin t/cli t/config-center-yaml t/control t/core t/debug t/discovery t/error_page t/misc
-          - t/node t/router t/script t/stream-node t/utils t/wasm
+          - t/node t/router t/script t/stream-node t/utils t/wasm t/amesh-library
```
Comment (spacewander) [inline .github/workflows/build.yml:32]: Should be sorted in order

File: apisix/cli/ops.lua [hunk 8]
```diff
@@ -560,6 +560,7 @@ Please modify "admin_key" in conf/config.yaml .
     end
     sys_conf["wasm"] = yaml_conf.wasm
 
+    sys_conf["config_center"] = yaml_conf.apisix.config_center
```
Comment (spacewander) [inline apisix/cli/ops.lua:563]: Handled in https://github.com/apache/apisix/blob/15517feb55a0c5e87c5fdb3b91d647ee0ca1c96f/apisix/cli/ops.lua#L555

File: apisix/cli/schema.lua [hunk 9]
```diff
@@ -28,7 +28,7 @@ local config_schema = {
         apisix = {
             properties = {
                 config_center = {
-                    enum = {"etcd", "yaml"},
+                    enum = {"etcd", "yaml", "shdict"},
```
Comment (spacewander) [inline apisix/cli/schema.lua:31]:
```suggestion
                    enum = {"etcd", "yaml", "xds"},
```
would be better?

File: apisix/core/config_shdict.lua [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        tried_paths[i] = fpath
+        i = i + 1
+    end
+
+    return nil, tried_paths
+end
+
+
+local function load_libamesh(lib_name)
+    ngx_timer_at(0, function(premature)
```
Comment (spacewander) [inline apisix/core/config_shdict.lua:82]: Why use the timer?

File: apisix/core/config_shdict.lua [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    if process.type() == "privileged agent" then
+        load_libamesh(lib_name)
+    end
+
+    return true
+end
+
+
+function _M.new(key, opts)
```
Comment (spacewander) [inline apisix/core/config_shdict.lua:115]: We can use injection for test?
https://github.com/apache/apisix/blob/15517feb55a0c5e87c5fdb3b91d647ee0ca1c96f/t/plugin/skywalking.t#L32-L50

File: .github/workflows/build.yml (hunk 7 above)
Comment (tzssangglass) [reply .github/workflows/build.yml]: fixed

File: apisix/cli/ops.lua (hunk 8 above)
Comment (tzssangglass) [reply apisix/cli/ops.lua]: rm this

File: apisix/cli/schema.lua (hunk 9 above)
Comment (tzssangglass) [reply apisix/cli/schema.lua]: updated

File: apisix/core/config_shdict.lua (hunk 10 above)
Comment (tzssangglass) [reply apisix/core/config_shdict.lua]: fixed

File: apisix/core/config_shdict.lua (hunk 11 above)
Comment (tzssangglass) [reply apisix/core/config_shdict.lua]: fixed

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -29,7 +29,7 @@ jobs:
         test_dir:
           - t/plugin
           - t/admin t/cli t/config-center-yaml t/control t/core t/debug t/discovery t/error_page t/misc
-          - t/node t/router t/script t/stream-node t/utils t/wasm
+          - t/amesh-library t/node t/router t/script t/stream-node t/utils t/wasm
```
Comment (tokers) [inline .github/workflows/build.yml:32]: What is amesh?

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -90,6 +90,11 @@ jobs:
           sudo dpkg -i tinygo_${TINYGO_VER}_amd64.deb
           cd t/wasm && find . -type f -name "*.go" | xargs -Ip tinygo build -o p.wasm -scheduler=none -target=wasi p
 
+      - name: Build Amesh library
```
Comment (tokers) [inline .github/workflows/build.yml:93]: Ditto.

File: apisix/core/config_xds.lua [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    local_conf = config_local.local_conf,
+}
+
+
+-- todo: refactor this function in chash.lua and radixtree.lua
+local function load_shared_lib(lib_name)
+    local string_gmatch = string.gmatch
+    local string_match = string.match
+    local io_open = io.open
+    local io_close = io.close
```
Comment (tokers) [inline apisix/core/config_xds.lua:58]: Put them to the module level?

File: apisix/core/config_xds.lua [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+    local cpath = package.cpath
+    local tried_paths = new_tab(32, 0)
+    local i = 1
+
+    for k, _ in string_gmatch(cpath, "[^;]+") do
```
Comment (tokers) [inline apisix/core/config_xds.lua:64]: Why not use `ngx.re.gmatch`?

File: apisix/core/config_xds.lua [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        local fpath = string_match(k, "(.*/)")
```
Comment (tokers) [inline apisix/core/config_xds.lua:65]: What about `ngx.re.match`?

File: apisix/init.lua [hunk 17]
```diff
@@ -110,6 +110,11 @@ function _M.http_init_worker()
     end
     require("apisix.balancer").init_worker()
     load_balancer = require("apisix.balancer")
+
+    if core.config == require("apisix.core.config_shdict") then
```
Comment (tokers) [inline apisix/init.lua:114]: I think we may need to unify the term, the shdict is the container, the xds is the way, I think here `config_xds` will be better.

Comment (tokers) [reply apisix/init.lua]: @membphis I don't think so, from the point of view of Apache APISIX, what is `amesh`? Why do we need to care about a commercial product?

File: apisix/core/config_xds.lua (hunk 15 above)
Comment (tzssangglass) [reply apisix/core/config_xds.lua]: updated

File: apisix/core/config_xds.lua (hunk 14 above)
Comment (tzssangglass) [reply apisix/core/config_xds.lua]: updated

File: .github/workflows/build.yml (hunk 13 above)
Comment (tzssangglass) [reply .github/workflows/build.yml]: updated

File: .github/workflows/build.yml (hunk 12 above)
Comment (tzssangglass) [reply .github/workflows/build.yml]: fixed, use `xds`

**Review by membphis — APPROVED** [review]
I think we can merge this MVP version. Then we can continue do other things

**Review by tao12345666333 — COMMENTED** [review]
In my opinion, I think all xds should be replaced with `external` or other names of similar meaning. 

Although we are currently integrating Istio/xds, APISIX does not actually need to pay attention to whether the external Go application is docking with xDS or other protocols.  If we name it `external` library, then users can use this mode to interface with any service, and even we can use it to watch etcd directly, it will be a very big change, and it will be more clear.

**Comment by spacewander** [comment]
IMHO, `external` is too general to know what it does. As etcd is an external service, what is the difference between `etcd` and `external`?
Moreover, we don't promise any API spec about the xds exchange format and we might optimize it for xds in the future. So I think using the name xds is specific enough for this feature.

File: t/xds-library/config_xds.t [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+    $block->set_value("extra_init_by_lua", $extra_init_by_lua);
+});
+
+run_tests;
+
+__DATA__
+
+=== TEST 1: load Amesh library so successfully
+--- yaml_config
```
Comment (spacewander) [inline t/xds-library/config_xds.t:69]: Let's set the yaml_config in the file level?

File: apisix/init.lua [hunk 19]
```diff
@@ -110,6 +110,11 @@ function _M.http_init_worker()
     end
     require("apisix.balancer").init_worker()
     load_balancer = require("apisix.balancer")
+
+    if core.config == require("apisix.core.config_xds") then
```
Comment (spacewander) [inline apisix/init.lua:114]: We can call it like:
https://github.com/apache/apisix/blob/9d450d7fe3169a77727df28696d083809b93977a/apisix/init.lua#L83

**Comment by tao12345666333** [comment]
> So I think using the name xds is specific enough for this feature.

OK so we can keep this naming for now and let's move forward.

File: apisix/init.lua (hunk 19 above)
Comment (tzssangglass) [reply apisix/init.lua]: updated

File: t/xds-library/config_xds.t (hunk 18 above)
Comment (tzssangglass) [reply t/xds-library/config_xds.t]: updated

**Comment by membphis** [comment]
we can merge this PR after fixed CI problems

File: apisix/cli/ngx_tpl.lua [hunk 20]
```diff
@@ -235,6 +235,10 @@ http {
     lua_shared_dict ext-plugin {* http.lua_shared_dict["ext-plugin"] *}; # cache for ext-plugin
     {% end %}
 
+    {% if config_center == "xds" then %}
+    lua_shared_dict router-config  10m;
```
Comment (spacewander) [inline apisix/cli/ngx_tpl.lua:239]:
```suggestion
    lua_shared_dict xds-route-config  10m;
```
would that be better?

File: apisix/core/config_xds.lua [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+local error             = error
+local is_http           = ngx.config.subsystem == "http"
+local io                = io
+local io_open           = io.open
+local io_close          = io.close
+local package           = package
+local new_tab           = base.new_tab
+local ffi               = require ("ffi")
+local C                 = ffi.C
+local router_config     = ngx.shared["router-config"]
```
Comment (spacewander) [inline apisix/core/config_xds.lua:34]: It is not configure for router, but for route.

File: apisix/core/config_xds.lua [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+    return nil, tried_paths
+end
+
+
+local function load_libxds(lib_name)
+    local xdsagent, tried_paths = load_shared_lib(lib_name)
+
+    if not xdsagent then
+        tried_paths[#tried_paths + 1] = 'tried above paths but can not load ' .. lib_name
+        error("can not load Amesh library, tried paths: " ..
```
Comment (spacewander) [inline apisix/core/config_xds.lua:100]:
```suggestion
        error("can not load xds library, tried paths: " ..
```

File: apisix/core/config_xds.lua (hunk 21 above)
Comment (tzssangglass) [reply apisix/core/config_xds.lua]: updated

File: apisix/cli/ngx_tpl.lua (hunk 20 above)
Comment (tzssangglass) [reply apisix/cli/ngx_tpl.lua]: updated

File: apisix/core/config_xds.lua (hunk 22 above)
Comment (tzssangglass) [reply apisix/core/config_xds.lua]: updated

File: apisix/core/config_xds.lua [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+local error             = error
+local is_http           = ngx.config.subsystem == "http"
+local io                = io
+local io_open           = io.open
+local io_close          = io.close
+local package           = package
+local new_tab           = base.new_tab
+local ffi               = require ("ffi")
+local C                 = ffi.C
+local router_config     = ngx.shared["xds-route-config"]
```
Comment (spacewander) [inline apisix/core/config_xds.lua:34]: There are still using `router` in some places.

**Review by membphis — COMMENTED** [review]
LGTM

File: apisix/core/config_xds.lua (hunk 23 above)
Comment (tzssangglass) [reply apisix/core/config_xds.lua]: updated

**Review by spacewander — APPROVED** [review]

File: apisix/core/config_xds.lua [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+local error             = error
+local is_http           = ngx.config.subsystem == "http"
+local io                = io
+local io_open           = io.open
+local io_close          = io.close
+local package           = package
+local new_tab           = base.new_tab
+local ffi               = require ("ffi")
+local C                 = ffi.C
+local route_config     = ngx.shared["xds-route-config"]
```
Comment (leslie-tsang) [inline apisix/core/config_xds.lua:34]:
```suggestion
local route_config      = ngx.shared["xds-route-config"]
```

File: t/APISIX.pm [hunk 25]
```diff
@@ -509,6 +509,7 @@ _EOC_
     lua_shared_dict ext-plugin 1m;
     lua_shared_dict kubernetes 1m;
     lua_shared_dict tars 1m;
+    lua_shared_dict xds-route-config 1m;
```
Comment (leslie-tsang) [inline t/APISIX.pm:512]: Should we sync the setting in `apisix/cli/ngx_tpl.lua` ?

Change `1m` to `10m`

Comment (tzssangglass) [reply t/APISIX.pm:512]: It is enough in the test.

File: apisix/core/config_xds.lua (hunk 24 above)
Comment (tzssangglass) [reply apisix/core/config_xds.lua:34]: I want to change it in the next PR😅, otherwise it will dismiss the previous approval.

Comment (leslie-tsang) [reply apisix/core/config_xds.lua:34]: OK

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**